internal/git/gitsync_processor.go

```diff
@@ -76,56 +78,107 @@ func watchRepo(ctx context.Context, restConfig *rest.Config, repo *v1.Repository
-	if err != nil {
-		logger.Errorw("error get the commit tree", "err", err)
-		return err
+	gitSync = &v1.GitSync{}
+	if err = k8Client.Get(ctx, namespacedName, gitSync); err != nil {
+		// if we aren't able to do a Get, then either it's been deleted in the past, or something else went wrong
+		if apierrors.IsNotFound(err) {
+			return nil
+		} else {
+			logger.Errorw("Unable to get GitSync", "err", err)
+			return err
+		}
 	}
+	val, ok := gitSync.Status.CommitStatus[repo.Name]
+	// Only create the resources for the first time if not created yet.
+	// Otherwise, monitoring with intervals.
+	if !ok {
+		// Retrieving the commit object matching the hash.
+		commit, err := r.CommitObject(*hash)
+		if err != nil {
+			logger.Errorw("error checkout the commit", "hash", hash.String(), "err", err)
+			return err
+		}
 
-	if !isRootDir(repo.Path) {
-		// Locate the tree with the given path.
-		tree, err = tree.Tree(repo.Path)
+		// Retrieve the tree from the commit.
+		tree, err := commit.Tree()
 		if err != nil {
-			logger.Errorw("error locate the path", "err", err)
+			logger.Errorw("error get the commit tree", "err", err)
 			return err
 		}
-	}
 
-	// Read all the files under the path and apply each one respectively.
-	err = tree.Files().ForEach(func(f *object.File) error {
-		logger.Debugw("read file", "file_name", f.Name)
-		//TODO: this currently assumes that one file contains just one manifest - modify for multiple
-		manifest, err := f.Contents()
+		if !isRootDir(repo.Path) {
+			// Locate the tree with the given path.
+			tree, err = tree.Tree(repo.Path)
+			if err != nil {
+				logger.Errorw("error locate the path", "repository path", repo.Path, "err", err)
+				return err
+			}
+		}
+
+		// Read all the files under the path and apply each one respectively.
+		err = tree.Files().ForEach(func(f *object.File) error {
+			logger.Debugw("read file", "file_name", f.Name)
+			// TODO: this currently assumes that one file contains just one manifest - modify for multiple
+			manifest, err := f.Contents()
+			if err != nil {
+				logger.Errorw("cannot get file content", "filename", f.Name, "err", err)
+				return err
+			}
+
+			// Apply manifest in cluster
+			return client.ApplyResource([]byte(manifest), namespace)
+		})
 		if err != nil {
-			logger.Errorw("cannot get file content", "filename", f.Name, "err", err)
 			return err
 		}
 
-		// Apply manifest in cluster
-		return k8sClient.ApplyResource([]byte(manifest), namespace)
-	})
-	if err != nil {
-		return err
+		commitStatus := v1.CommitStatus{
+			Hash:     hash.String(),
+			Synced:   true,
+			SyncTime: metav1.NewTime(time.Now()),
+		}
+
+		gitSync = &v1.GitSync{}
+		if err = k8Client.Get(ctx, namespacedName, gitSync); err != nil {
+			// if we aren't able to do a Get, then either it's been deleted in the past, or something else went wrong
+			if apierrors.IsNotFound(err) {
+				return nil
+			} else {
+				logger.Errorw("Unable to get GitSync", "err", err)
+				return err
+			}
+		}
```

Comment:
why do we do the `Get()` again if we just did it? Maybe we should do the `Update()` and then if it fails, repeatedly do `Get()` and then `Update()` in an exponential backoff? And put the whole thing in a method so that @shubhamdixit863 will be able to call it to?

Reply:
Add a method for update commit status specifically. For exponential backoff, we discussed offline, it would be address in https://github.com/numaproj-labs/numaplane/issues/46 as we are switching to work queue model.